### ISSUE TITLE
Table's distribution characteristics declarative, in CREATE TABLE

### ISSUE THREAD

**Issue by robin900** [citusdata/citus#852]
I'm wondering if there is or has been a desire to make distributed-ness of tables declarative; that is, part of the CREATE TABLE statement and processed as part of the CREATE TABLE operation. 

If something extra in the table definition could be used to indicate distribution method, partition column, shard count, and replication factor (and maybe the colocation relationships when those are added), then the multi-planner could extract those values from the "extra thing"; then it would discard the extra thing from the table definition, and proceed with a standard table create followed by the operations to create_distributed_table() and create_worker_shards(). CREATE TABLE statements that lack this extra thing would result in an error (if only distributed non-temp tables are allowed in Citus).

In PG <= 9.5 the only possibility I can imagine is to involve a dummy column of a new data type `citus.distributed('hash', partition_col, 8, 1)`, which could store all of the needed info in the typmod int32 of the data type. CREATE TABLE statements that lacked a column with type citus.distributed would result in an error. But the typmod in32 is too small to be able to include a usable reference to the partition column (only available as a cstring and no lookup on the table definition is possible) plus the other values. So <= 9.5 it doesn't seem possible.

In PG >= 9.6, Citus could define a new `ACCESS METHOD` whose entire job is to allow `CREATE TABLE` statements to include `INDEX (partition_col) USING distributed WITH (method='hash', shard_count=8, replication_factor=1)`. The index itself wouldn't be created, and no standalone `CREATE INDEX ... USING distributed` would be allowed by the multi-planner. The access method only exists to allow the configuration values ("storage parameters" in this case) to reach the multi-planner; the index definition is then discarded. Or, alternately, the index *is* created so that it shows up via reflection in pg_catalog metadata, but does not participate in query plans because of the access method implementation.

**Comment by anarazel** [comment]
When doing the work of turning citus in an extension, I was actually
proposing that, and had initially implemented it.  What I used there was
to specify the partitioning info using
```CREATE TABLE .. WITH(citus.partition_method = '...')```

We decided to not go that way.

**Comment by robin900** [comment]
I did not know that table storage parameters were "free form" such that one could specify any name=value pair one likes. Perhaps it's permissible in the parse tree but not in the plan tree? I can imagine the bookkeeping necessary to capture the dist config from the parse tree and preserve it through the standard planner into a plan tree, for processing by the multi planner, might be too much.

**Comment by amosbird** [comment]
@robin900 I've been thinking similar things. What I do now is creating an event trigger.

```
CREATE OR REPLACE FUNCTION trigger_table_distribution()
    RETURNS event_trigger
    LANGUAGE plpgsql
    AS $$
DECLARE
    name TEXT;
    node_list TEXT;
    key  TEXT;
    first_key  TEXT;
    distribution_type citus.distribution_type;
    shard_num INTEGER;
    rep_num INTEGER;
BEGIN
    SELECT object_identity INTO name FROM pg_event_trigger_ddl_commands()
    WHERE schema_name NOT LIKE 'pg_%';
    SELECT master_get_active_worker_nodes() INTO node_list;
    IF name IS NOT NULL AND node_list IS NOT NULL THEN
        SELECT a.attname INTO key
        FROM   pg_index i
        JOIN   pg_attribute a ON a.attrelid = i.indrelid
        AND    a.attnum = ANY(i.indkey)
        WHERE  i.indrelid = name::regclass
        AND    i.indisprimary;
        SELECT attname INTO first_key FROM pg_attribute
        WHERE attrelid = name::regclass
        AND attnum > 0 AND NOT attisdropped ORDER BY attnum LIMIT 1;
        SELECT current_setting('citus.distribution_type')::citus.distribution_type INTO distribution_type;
        SELECT current_setting('citus.shard_num')::int INTO shard_num;
        SELECT current_setting('citus.shard_replication_factor')::int INTO rep_num;
        IF key IS NOT NULL THEN
            IF distribution_type = 'append' THEN
                raise EXCEPTION 'cannot use append distribution when table has primary key.';
            END IF;
            RAISE NOTICE 'use column % as distributed key.', key;
            PERFORM master_create_distributed_table(name, key, distribution_type);
            PERFORM master_create_worker_shards(name, shard_num, rep_num);
        ELSE
            RAISE WARNING 'without primary key. implicitly use column % as distributed key.', first_key;
            PERFORM master_create_distributed_table(name::regclass, first_key, distribution_type);
            IF distribution_type = 'hash' THEN
                PERFORM master_create_worker_shards(name, shard_num, rep_num);
            END IF;
        END IF;
    END IF;
END;
$$;

COMMENT ON FUNCTION trigger_table_distribution()
    IS 'automatically create distributed table using ddl event trigger.';

CREATE EVENT TRIGGER trigger_table_distribution
    ON ddl_command_end
    WHEN TAG in ('CREATE TABLE')
    EXECUTE PROCEDURE trigger_table_distribution();
```

Hope it may help ;)

**Comment by marcocitus** [comment]
It would be nicer for MX, since that would allow you to create a distributed table from any node. If you do a regular CREATE TABLE on a worker it would just be on that node and you might get connected to a different one before you run master_create_distributed_table. (Obviously, we also need the actual logic to propagate the table creation from a worker)

**Comment by robin900** [comment]
@amosbird Thanks for the example code!

**Comment by ozgune** [comment]
@amosbird -- this is interesting. Could you elaborate a bit more on your motivation for doing this? In your use case, were all tables defined as distributed tables and you wanted to simplify the ``CREATE TABLE`` calls?

**Comment by amosbird** [comment]
@ozgune yes, I'd like current users to use distributed tables without changing their DDLs.
